gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/MessageConsumerImpl.java

```diff
@@ -0,0 +1,263 @@
+
+    private final ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    @Override
+    public ExecutorService get() {
+      return executor;
+    }
+
+    @Override
+    public void release(ExecutorService executor) {
+      executor.shutdownNow();
+    }
+  }
+
+  class ConsumerRunnable implements Runnable {
+
+    @Override
+    public void run() {
+      if (closed) {
+        return;
+      }
+      pullerFuture = pubsubRpc.pull(createPullRequest());
+      pullerFuture.addCallback(new PullCallback() {
+        @Override
+        public void success(PullResponse response) {
+          List<com.google.pubsub.v1.ReceivedMessage> messages = response.getReceivedMessagesList();
+          queuedCallbacks.addAndGet(messages.size());
+          for (com.google.pubsub.v1.ReceivedMessage message : messages) {
+            deadlineRenewer.add(subscription, message.getAckId());
+            ReceivedMessage receivedMessage = ReceivedMessage.fromPb(pubsub, subscription, message);
+            executor.execute(ackingRunnable(receivedMessage));
+          }
+          nextPull();
+        }
+
+        @Override
+        public void failure(Throwable error) {
+          if (!(error instanceof CancellationException)) {
+            nextPull();
+          }
+        }
+      });
+    }
+  }
+
+  private MessageConsumerImpl(Builder builder) {
+    this.pubsubOptions = builder.pubsubOptions;
+    this.subscription = builder.subscription;
+    this.messageProcessor = builder.messageProcessor;
+    this.pubsubRpc = pubsubOptions.rpc();
+    this.pubsub = pubsubOptions.service();
+    this.deadlineRenewer = builder.deadlineRenewer;
+    this.queuedCallbacks = new AtomicInteger();
+    this.timer = SharedResourceHolder.get(TIMER);
+    this.executorFactory = firstNonNull(builder.executorFactory, new DefaultExecutorFactory());
+    this.executor = executorFactory.get();
+    this.maxQueuedCallbacks = firstNonNull(builder.maxQueuedCallbacks, MAX_QUEUED_CALLBACKS);
+    this.consumerRunnable = new ConsumerRunnable();
+    nextPull();
+  }
+
+  private Runnable ackingRunnable(final ReceivedMessage receivedMessage) {
+    return new Runnable() {
+      @Override
+      public void run() {
+        try {
+          messageProcessor.process(receivedMessage);
+          pubsub.ackAsync(receivedMessage.subscription(), receivedMessage.ackId());
+        } catch (Exception ex) {
+          pubsub.nackAsync(receivedMessage.subscription(), receivedMessage.ackId());
+        } finally {
+          deadlineRenewer.remove(receivedMessage.subscription(), receivedMessage.ackId());
+          queuedCallbacks.decrementAndGet();
+          // We can now pull more messages. We do not pull immediately to possibly wait for other
+          // callbacks to end
+          scheduleNextPull(500, TimeUnit.MILLISECONDS);
+        }
+      }
+    };
+  }
+
+  private PullRequest createPullRequest() {
```

Comment:
Some methods such as `createPullRequest` and `ackingRunnable` could be placed closer to their context (e.g. inner class).